### PR TITLE
FUN-5282-Create-SubCanvas-for-each-Popups-and-Windows

### PR DESCRIPTION
**Pull request by alexveligurskiy**

### CODE REVIEW

File: Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs [hunk 1]
```diff
@@ -52,22 +58,36 @@ public class SimpleFadeTransition : ATransitionComponent
...
+            else
+            {
+                if (fadeOut)
+                {
+                    canvasGroup.alpha = 0f;
+                }
+                else
+                {
+                    canvasGroup.alpha = 1f;
+                }
```
Comment (TarasMartynyuk) [inline Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs:87]: ternary operator?

File: Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs [hunk 2]
```diff
@@ -52,22 +58,36 @@ public class SimpleFadeTransition : ATransitionComponent
...
+
+                if (currentAction != null)
```
Comment (TarasMartynyuk) [inline Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs:89]: can use ?.Invoke here

File: .vsconfig [hunk 3]
```diff
@@ -0,0 +1,6 @@
+﻿{
+  "version": "1.0",
+  "components": [
+    "Microsoft.VisualStudio.Workload.ManagedGame"
+  ]
+}
```
Comment (TarasMartynyuk) [inline .vsconfig:6]: this is maybe some visual studio/code internal file?
probably should be ignored then

File: ProjectSettings/ProjectSettings.asset [hunk 4]
```diff
@@ -68,6 +68,12 @@ PlayerSettings:
   androidRenderOutsideSafeArea: 1
   androidUseSwappy: 1
   androidBlitType: 0
+  androidResizableWindow: 0
+  androidDefaultWindowWidth: 1920
+  androidDefaultWindowHeight: 1080
```
Comment (TarasMartynyuk) [inline ProjectSettings/ProjectSettings.asset:73]: fyi, our target resoulution is not 1080p, it's larger - 1125 or smth,
not sure if relevant here

File: Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs [hunk 5]
```diff
@@ -20,27 +20,33 @@ public class SimpleFadeTransition : ATransitionComponent
...
 
-        public override void Animate(Transform target, Action callWhenFinished) {
-            if (currentAction != null) {
-                canvasGroup.alpha = endValue;
-                currentAction();
-            }
-
+        public override void Animate(Transform target, Action callWhenFinished)
+        {
+            currentTarget = target;
```
Comment (lvgrecords) [inline Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs:30]: `currentTarget` is not used, let remove it

File: Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs [hunk 6]
```diff
@@ -20,27 +20,33 @@ public class SimpleFadeTransition : ATransitionComponent
         private Action currentAction;
         private Transform currentTarget;
 
-        private float startValue;
-        private float endValue;
+        private float startValue = 0f;
```
Comment (lvgrecords) [inline Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs:23]: what is the point set some value? it anyhow will be overwritten on `Animate()`

File: Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs [hunk 7]
```diff
@@ -52,22 +58,36 @@ public class SimpleFadeTransition : ATransitionComponent
...
+        public override void Stop(Transform target)
+        {
             shouldAnimate = false;
         }
 
-        private void Update() {
-            if (!shouldAnimate) {
+        private void Update()
+        {
+            if (!shouldAnimate || canvasGroup == null)
```
Comment (lvgrecords) [inline Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs:68]: is not `canvasGroup == null` only if `shouldAnimate == false`?

File: Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs [hunk 8]
```diff
@@ -20,27 +20,33 @@ public class SimpleFadeTransition : ATransitionComponent
...
+            currentTarget = target;
             canvasGroup = target.GetComponent<CanvasGroup>();
-            if (canvasGroup == null) {
+            if (canvasGroup == null)
+            {
                 canvasGroup = target.gameObject.AddComponent<CanvasGroup>();
             }
 
-            if (fadeOut) {
+            if (currentAction != null)
```
Comment (lvgrecords) [inline Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs:37]: looks previous impl with calling `currentAction()` (actually it's previous action) before setting `canvasGroup` value correct, it works aka set final state for previous `canvasGroup`  if we call `Animate()` while previous animate not finished

what is the point to change it?

File: Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs (hunk 1 above)
Comment (lvgrecords) [reply Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs:87]: or `canvasGroup.alpha = endValue`

File: .vsconfig (hunk 3 above)
Comment (alexveligurskiy) [reply .vsconfig:6]: done

File: Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs (hunk 6 above)
Comment (alexveligurskiy) [reply Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs:23]: no point, just tested some things and didnt removed

File: Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs (hunk 5 above)
Comment (alexveligurskiy) [reply Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs:30]: done

File: Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs (hunk 8 above)
Comment (alexveligurskiy) [reply Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs:37]: what would be if we call this method for the first time without action and the canvas group is not exists there?

File: Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs (hunk 7 above)
Comment (alexveligurskiy) [reply Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs:68]: agree

File: Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs (hunk 1 above)
Comment (alexveligurskiy) [reply Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs:87]: agree with endvalue

File: Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs (hunk 2 above)
Comment (alexveligurskiy) [reply Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs:89]: agree, it is much more readable

File: Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs (hunk 8 above)
Comment (lvgrecords) [reply Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs:35]: not sure I get it, so pls correct me if I wrong/miss something - `Animate()` called by framework internally and guarantee that action will be always passed

Comment (alexveligurskiy) [reply Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs:35]: for now, we can guarantee that AUIScreenController would always pass OnTransitionInFinished/OnTransitionOutFinished actions, but, if there would be created some kind of additional implementation of AUIScreenController without Actions or for some reason canvasgroup would be destroyed - we would run into a problem here

Comment (lvgrecords) [reply Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs:35]: mb, so assume we can fix when (if) it happens

but it do not decline fact that right now we broke consistency - previously saved action could be called with new canvas group

Comment (alexveligurskiy) [reply Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs:35]: why there should be a new canvas group? we always push AUIScreenController transform as "target"

Comment (alexveligurskiy) [reply Assets/Runtime/ScreenTransitions/SimpleFadeTransition.cs:35]: yep, I agree that it should be on awake, can do that - if we are not going to call Animate() from screens that are not on the same gameobject

**Pull request merged**